File: frontend/src/types/homework.ts
export interface HomeworkAssignment {
  id: string;
  tutor_id: string;
  student_id: string | null;
  group_id: string | null;
  title: string | null;
  content_type: string;
  content_url: string | null;
  content_file_path: string | null;
  submission_mode: string;
  due_at: string | null;
  created_at: string;
  status: string;
  student_display_name: string | null;
  group_name: string | null;
  // Сдачи учеников приходят вместе с заданием - карточка репетитора показывает их сразу.
  submissions: HomeworkSubmission[];
}

export interface HomeworkSubmissionFile {
  id: string;
  file_path: string;
  uploaded_at: string;
}

export interface HomeworkSubmission {
  id: string;
  assignment_id: string;
  student_id: string;
  status: string;
  // Файлов может быть несколько: ученик добавляет ещё один скриншот или убирает лишний.
  files: HomeworkSubmissionFile[];
  comment: string | null;
  submitted_at: string | null;
  student_display_name: string | null;
}

export interface StudentHomework {
  submission_id: string;
  assignment_id: string;
  tutor_id: string;
  group_id: string | null;
  // Может быть пустым: задание бывает одной ссылкой, без названия.
  title: string | null;
  content_type: string;
  content_url: string | null;
  content_file_path: string | null;
  submission_mode: string;
  due_at: string | null;
  status: string;
  files: HomeworkSubmissionFile[];
  comment: string | null;
  submitted_at: string | null;
}

File: backend/app/schemas/homework.py
import uuid

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.common import UTCDateTime


class HomeworkAssignmentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    tutor_id: uuid.UUID
    student_id: uuid.UUID | None
    group_id: uuid.UUID | None
    title: str | None
    content_type: str
    content_url: str | None
    content_file_path: str | None
    submission_mode: str
    due_at: UTCDateTime | None
    created_at: UTCDateTime
    # Populated by list_my_assignments (api/v1/homework.py) - not meaningful right
    # after create/update, which return before there's anything to aggregate/name.
    status: str = "pending"
    student_display_name: str | None = None
    group_name: str | None = None
    # Сдачи учеников прямо в карточке: репетитору нужно видеть, кто что прислал, не
    # раскрывая каждое задание отдельно.
    submissions: list["HomeworkSubmissionOut"] = Field(default_factory=list)


class HomeworkSubmissionFileOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    file_path: str
    uploaded_at: UTCDateTime


class HomeworkSubmissionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    assignment_id: uuid.UUID
    student_id: uuid.UUID
    status: str
    # Файлов может быть несколько: ученик прикладывает скриншот, при необходимости
    # добавляет ещё один, а ошибочный убирает.
    files: list[HomeworkSubmissionFileOut] = Field(default_factory=list)
    comment: str | None
    submitted_at: UTCDateTime | None
    # Проставляется списком сдач: репетитору нужно имя, а не идентификатор.
    student_display_name: str | None = None


HomeworkAssignmentOut.model_rebuild()


class HomeworkSubmissionStatusUpdate(BaseModel):
    status: str


class StudentHomeworkOut(BaseModel):
    submission_id: uuid.UUID
    assignment_id: uuid.UUID
    tutor_id: uuid.UUID
    group_id: uuid.UUID | None
    # Название необязательно с миграции afc6a11a104d: репетитор может задать домашку
    # одной ссылкой. Здесь оно оставалось обязательным, и список домашних заданий у
    # такого ученика падал с 500 вместо того, чтобы показать задание без заголовка.
    title: str | None
    content_type: str
    content_url: str | None
    content_file_path: str | None
    submission_mode: str
    due_at: UTCDateTime | None
    status: str
    files: list[HomeworkSubmissionFileOut] = Field(default_factory=list)
    comment: str | None
    submitted_at: UTCDateTime | None
